### PR TITLE
[perf] Feature: Adding support for async profiling

### PR DESCRIPTION
**Pull request by iyzhang**
This PR adds a feature to our profiler to measure the time spent in async blocks. It only supports fused (non-reentrant) blocks for now. It works by simply wrapping a Future around the existing code that measures when we enter the async block and stops the clock when we exit with either a Pending or Ready status. 

For now, I have annotated all of our coroutines but nothing more detailed. We should add more annotations, especially around the different async workers that all run in a single background coroutine.

### CODE REVIEW

File: src/rust/perftools/profiler/mod.rs [hunk 1]
```diff
@@ -63,11 +41,13 @@ macro_rules! timer {
 /// root nodes. Thus, if you have multiple root nodes and they do not cover
 /// all code that runs in your program, the printed frequencies will be
 /// overestimated.
+
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:44]: Please rollback this change.

File: src/rust/perftools/profiler/mod.rs [hunk 2]
```diff
@@ -63,11 +41,13 @@ macro_rules! timer {
...
 pub fn write<W: io::Write>(out: &mut W, max_depth: Option<usize>) -> io::Result<()> {
     PROFILER.with(|p| p.borrow().write(out, max_depth))
 }
 
 /// Reset profiling information.
+
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:50]: Please rollback this change.

File: src/rust/perftools/profiler/mod.rs [hunk 3]
```diff
@@ -203,6 +229,7 @@ impl Drop for Guard {
 /// Note that there is a global thread-local instance of `Profiler` in
 /// [`PROFILER`](constant.PROFILER.html), so it is not possible to manually
 /// create an instance of `Profiler`.
+
```
Comment (ppenna) [inline src/rust/perftools/profiler/mod.rs:232]: Please rollback this change.

**Review by ppenna — APPROVED** [review]
LGTM

**Comment by iyzhang** [comment]
Note the statistics at the end of this file:
[tcp-echo-sequential-1-64-128-client-demikernel-vm5.eastus.cloudapp.azure.com.stdout.txt](https://github.com/microsoft/demikernel/files/14641466/tcp-echo-sequential-1-64-128-client-demikernel-vm5.eastus.cloudapp.azure.com.stdout.txt)

**Pull request merged**